src/control/system/membership.go

```diff
@@ -662,37 +662,49 @@ func (m *Membership) IsRankAdminExcluded(rank Rank) bool {
 	return cm.State == MemberStateAdminExcluded
 }
 
-// MarkRankDead is a helper method to mark a rank as dead in response to a
-// swim_rank_dead event.
-func (m *Membership) MarkRankDead(rank Rank, incarnation uint64) error {
+// MarkRankDead is a helper method to mark a rank as dead in response to a swim_rank_dead event.
+// Pass zero value for incarnation to skip comparison with existing member incarnation value.
+// Return bool indicates whether a membership change was made and might indicate a group update is
+// necessary.
+func (m *Membership) MarkRankDead(rank Rank, incarnation uint64) (bool, error) {
 	m.Lock()
 	defer m.Unlock()
 
 	member, err := m.db.FindMemberByRank(rank)
 	if err != nil {
-		return err
+		return false, err
 	}
 
 	ns := MemberStateExcluded
+	msg := MsgBadStateTransition(member, ns)
+	if member.State == ns {
+		m.log.Trace(msg)
+		return false, nil
+	}
 	if member.State.isTransitionIllegal(ns) {
-		msg := msgBadStateTransition(member, ns)
-		// excluded->excluded transitions expected for multiple swim
-		// notifications, if so return error to skip group update
-		if member.State != ns {
-			m.log.Error(msg)
-		}
-
-		return errors.New(msg)
+		// This doesn't constitute an error, just that a transition from the current state
+		// to excluded isn't allowed. Possibly the engine is in start-up and is no longer
+		// in a joined state.
+		m.log.Debug(msg)
+		return false, nil
 	}
 
-	if member.State == MemberStateJoined && member.Incarnation > incarnation {
-		m.log.Debugf("ignoring rank dead event for previous incarnation of %d (%x < %x)", rank, incarnation, member.Incarnation)
-		return errors.Errorf("event is for previous incarnation of %d", rank)
+	// Perform check on rank incarnation number if valid value passed and a currently
+	// active/joined rank is going to be excluded.
+	if member.State == MemberStateJoined && incarnation > 0 &&
```

Comment:
I don't love using a magic incarnation value as an indirect "force" flag.

Reply:
Fine with changing as a follow-up. I'd suggest just a "force" if that's what's intended, i.e. always send the event, don't bother checking anything.

Reply:
I don't want to skip verifying against illegal transitions though, should I use a forceIncarnation flag or just leave it?